### PR TITLE
ENH: empty and nan input handling for `stats.kstat` and `stats.kstatvar`

### LINKED ISSUE #675: Statistics Review: kstat (Trac #148)
_Original ticket http://projects.scipy.org/scipy/ticket/148 on 2006-04-02 by @rkern, assigned to @rkern._

The function `kstat` in file source:trunk/Lib/stats/morestats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### LINKED ISSUE #676: Statistics Review: kstatvar (Trac #149)
_Original ticket http://projects.scipy.org/scipy/ticket/149 on 2006-04-02 by @rkern, assigned to @rkern._

The function `kstatvar` in file source:trunk/Lib/stats/morestats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-675; closes gh-676

### CODE REVIEW

**Comment by aeklant** [comment]
Notes:
* nan input returned nan already but I made it explicit to be sure.
* The test for equivalence with `moment` could have been done in one step but the PR adding broadcastable input handling for it (gh-5000) hasn't been merged yet.
* I haven't added the example including both `kstat` and `kstatvar` applied to a distribution because I couldn't come up with a comprehensible, organic example. I would appreciate some help here.

**Comment by ev-br** [comment]
Not sure about kstatvar, but for `kstat` one cute example could be to use the fact that for a normal distribution all higher cumulants should be zero.

```
In [48]: rndm = np.random.RandomState(1234)

In [49]: for n in [2, 3, 4, 5]:
    data = rndm.normal(size=10**n)
    k = stats.kstat(data, 4)
    m = stats.moment(data, 4) - 3*stats.moment(data, 2)**2
    print n, k, m
   ....:     
2 1.22353383243 1.08167257454
3 -0.0838122446309 -0.0885402457453
4 -0.0629243692927 -0.0634734782219
5 -0.0191377033499 -0.0191966961729
```

Also while you're at it, I'd suggest writing the formula in the docstring in a proper TeX notation.

**Comment by rgommers** [comment]
The explanation of equivalence of cumulants in terms of central moments in the docstring seems to be plain wrong. Would be better to give formulas 8-11 from http://mathworld.wolfram.com/k-Statistic.html, and also test that those are correct. I've checked for ``n = 1, 2``:

    In [7]: stats.kstat(x, 1)
    Out[7]: 0.29394148970071882

    In [8]: x.mean()
    Out[8]: 0.29394148970071882

    In [9]: stats.kstat(x, 2)
    Out[9]: 0.75916813881595457

    In [10]: 10./9 * x.var()
    Out[10]: 0.75916813881595469

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -182,9 +182,11 @@ def kstat(data, n=2):
     See Also
     --------
     kstatvar: Returns an unbiased estimator of the variance of the k-statistic.
+    moment: Returns the n-th central moment about the mean for a sample.
 
     Notes
     -----
+    The input data gets raveled prior to calculating the cumulants.
```
Comment (rgommers) [inline scipy/stats/morestats.py]: I'd move this to the parameter description for ``data``. Something like ``Input data.  Note that n-D input gets flattened.``

Comment (rgommers) [inline scipy/stats/morestats.py]: (and if it stayed here, would have needed a blank line under it)

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -213,6 +215,15 @@ def kstat(data, n=2):
     S = zeros(n + 1, 'd')
     data = ravel(data)
     N = len(data)
```
Comment (rgommers) [inline scipy/stats/morestats.py]: Can you make this ``N = data.size``? That's a bit more idiomatic.

File: scipy/stats/morestats.py [hunk 3]
```diff
@@ -213,6 +215,15 @@ def kstat(data, n=2):
     S = zeros(n + 1, 'd')
```
Comment (rgommers) [inline scipy/stats/morestats.py]: ``'d'`` is a bit obscure, how about making this ``np.zeros(n + 1, np.float64)``

File: scipy/stats/morestats.py [hunk 4]
```diff
@@ -213,6 +215,15 @@ def kstat(data, n=2):
...
+    # raise ValueError on empty input
+    if N == 0:
+        raise ValueError("Data input must not be empty")
+
+    # on nan input, return nan without warning
+    if np.isnan(np.sum(data)):
+        return np.nan
+
     for k in range(1, n + 1):
         S[k] = sum(data**k, axis=0)
```
Comment (rgommers) [inline scipy/stats/morestats.py]: This ``sum`` is ugly, is actually ``np.sum`` which gets imported as ``from numpy import sum`` that shadows the builtin ``sum``. Would be good to fix that in the whole file.

File: scipy/stats/tests/test_morestats.py [hunk 5]
```diff
@@ -677,6 +675,21 @@ def test_moments_normal_distribution(self):
         expected = [0.011315, 1.017931, 0.05811052, 0.0754134]
         assert_allclose(moments, expected, rtol=1e-4)
 
+        # test equivalence with `stats.moment`
+        m1 = stats.moment(data, moment=1)
+        m2 = stats.moment(data, moment=2)
+        m3 = stats.moment(data, moment=3)
+        assert_allclose((m1, m2, m3), expected[:-1], atol=1e-1)
```
Comment (rgommers) [inline scipy/stats/tests/test_morestats.py]: This is not the actual equivalence (docstring is wrong as I remarked on earlier). That's why you need ``atol=0.1`` to make the test pass:


    In [16]: stats.kstat(x, 3)
    Out[16]: -0.1788808700120989

    In [17]: stats.moment(x, 3)
    Out[17]: -0.12879422640871122

**Comment by ev-br** [comment]
Re relation between cumulants and central moments: AFAICS, what's written in the docstring is correct for n>=2.  

As a data point, I'm glad cumulants and the relation to momenta are at least mentioned. Just saying "k-statistic" might be self-explanatory for statisticians, but definitely is not for me (FWIW).

**Comment by rgommers** [comment]
@ev-br I don't quite see it (for example https://github.com/scipy/scipy/pull/5067#discussion_r35690033). Can you demonstrate that correctness in code for ``n=3``?

**Comment by ev-br** [comment]
```
In [22]: rndm = np.random.RandomState(1234)

In [23]: for n in [2, 3, 4, 5, 6, 7]:
    x = rndm.normal(size=10**n)
    m, k = moment(x, 3), kstat(x, 3)
    print m, k, m-k
   ....:     
0.17923117332 0.184736315523 -0.00550514220259
-0.0253939676808 -0.0254703277233 7.63600425147e-05
-0.054684694683 -0.0547011039202 1.64092371539e-05
-7.4672724767e-05 -7.46749650009e-05 2.24023397193e-09
0.00326127202153 0.00326128180537 -9.78383880943e-09
-2.84056226496e-05 -2.84056311714e-05 8.52181322267e-12

In [24]: rndm = np.random.RandomState(1234)

In [25]: for n in [2, 3, 4, 5, 6, 7]:
    x = rndm.uniform(size=10**n)
    m, k = moment(x, 3), kstat(x, 3)
    print m, k, m-k
   ....:     
-0.0038555916577 -0.00397401737549 0.00011842571779
-0.000743885441888 -0.000746122316593 2.236874705e-06
-0.000119314362361 -0.000119350165024 3.58026625511e-08
-2.33000307132e-05 -2.33007297305e-05 6.99017277607e-10
1.07278858677e-05 1.07279180514e-05 -3.21836475283e-11
4.10084770514e-06 4.10084893447e-06 -1.22932474493e-12
```

Am I just hitting a lucky random seed?

**Comment by ev-br** [comment]
One more example, where the expected result is non-zero:

```
In [36]: for n in [2, 3, 4, 5, 6, 7]:
    x = stats.poisson.rvs(4, size=10**n, random_state=rndm)
    m, k = moment(x, 3), kstat(x, 3)
    print m, k, m-k
   ....:     
3.403014 3.50753865182 -0.104524651824
4.759773522 4.77408623253 -0.0143127105251
3.83308102963 3.83423122231 -0.00115019268207
4.0624757209 4.06259759802 -0.000121877115421
4.00658646119 4.00659848098 -1.20197874587e-05
3.99163208223 3.99163327972 -1.19748991478e-06
```

For Poisson distribution, the 3rd central moment should equal `lambda` (e.g., https://en.wikipedia.org/wiki/Cumulant, sec "Cumulants of some discrete probability distributions")

**Comment by ev-br** [comment]
@rgommers what is `x` in https://github.com/scipy/scipy/pull/5067#discussion_r35690033? I tried reproducing the seed from the test:

```
In [37]: np.random.seed(32149)

In [38]: x = np.random.randn(12345)

In [39]: moment(x, 3)
Out[39]: 0.058096401760671108

In [40]: kstat(x, 3)
Out[40]: 0.058110522631750268
```

**Comment by rgommers** [comment]
@ev-br not a lucky seed - I was using inputs with too small size. Can't reproduce it exactly anymore because I didn't use a random seed, but your example clearly shows that from n=100 or so you get good matches (decent ``rtol``).

In the example it would be good then to show that ``kstat`` and ``moment`` *converge* to the same number but aren't actually *identical* (this is how I read it at first).

File: scipy/stats/tests/test_morestats.py (hunk 5 above)
Comment (rgommers) [inline scipy/stats/tests/test_morestats.py]: So after discussion with @ev-br: this uses ``data.size=12345`` which is large enough to expect a decent match. 

``atol=1e-1`` is very little. Can you use ``atol=002`` (for first moment, which gives 0.0 for ``moment`` and 0.0113 for ``kstat``) and then tighten ``rtol`` close to as far as possible? Should be 1e-2 at least I expect.

**Comment by rgommers** [comment]
Maybe this example https://github.com/scipy/scipy/pull/5067#issuecomment-125733129 would be appropriate to add to the docstring.

**Comment by aeklant** [comment]
Please review latest changes. The implementation changes are very small but I suppose there is a lot to nitpick in the documentation.

**Comment by ev-br** [comment]
Re documentation: could you build the docs locally and see if it renders? (`python setup.py install; cd docs; make html-scipyorg`). I suspect you'd need to make the docstring a raw string, r""", shield underscores and asteristsk from Sphinx by double backticks, and copyedit the TeX markup.

File: scipy/stats/morestats.py [hunk 6]
```diff
@@ -182,23 +182,21 @@ def kstat(data, n=2):
...
+        k_{3} = \frac{ n^{2} } {(n-1) (n-2)} m_{3}
+        k_{4} = \frac{ n^{2} [(n + 1)m_{4} - 3(n - 1) m^2_{2}]} {(n-1) (n-2) (n-3)}
 
-        kappa_1 = mu_1
-        kappa_2 = mu_2
-        kappa_3 = mu_3
-        kappa_4 = mu_4 - 3*mu_2**2
-        kappa_5 = mu_5 - 10*mu_2 * mu_3
+    where ``\mu`` is the sample mean, ``m_{2}`` is the sample variance, and
+    ``m_{i}`` is the i-th sample central moment.
```
Comment (ev-br) [inline scipy/stats/morestats.py]: This mixes the code markup (double back ticks) and TeX, so better do :math:`\m_2` and so on.

File: scipy/stats/morestats.py [hunk 7]
```diff
@@ -206,15 +204,42 @@ def kstat(data, n=2):
 
     http://mathworld.wolfram.com/Cumulant.html
 
+    Examples
+    --------
+    >>> from scipy import stats
+    >>> rndm = np.random.RandomState(1234)
+
+    As sample size increases, n-th moment and n-th k-statistic converge to the
+    same number (although they aren't identical)
```
Comment (ev-br) [inline scipy/stats/morestats.py]: ... and this "same number" is zero for the normal distribution, even though the convergence is not very fast.

File: scipy/stats/morestats.py [hunk 8]
```diff
@@ -206,15 +204,42 @@ def kstat(data, n=2):
...
+    >>> for n in [2, 3, 4, 5, 6, 7]:
+    ...     x = rndm.normal(size=10**n)
+    ...     m, k = stats.moment(x, 3), stats.kstat(x, 3)
+    ...     print m, k, m-k
+    -0.63144956970249455, -0.65084474304524276, 0.019395173342748206
+    0.028228612555103411, 0.028313496417362614, -8.4883862259202414e-05
+    -0.045424541104455429, -0.04543817164718604, 1.3630542730611039e-05
+    7.5338208014728139e-05, 7.5340468213851364e-05, -2.2601991232247673e-09
+    0.0016645501352486995, 0.0016645551289107396, -4.9936620400824888e-09
+    -2.8779648300317205e-06, -2.8779656933968186e-06, 8.6336509815359533e-13
```
Comment (ev-br) [inline scipy/stats/morestats.py]: Nitpick: This, I think, will look nicer with some formatting of numbers: I'd replace the print with e.g. `print "%.3g  %.3g  %.3g" %(m, k, m-k)` or equivalent.

Comment (rgommers) [inline scipy/stats/morestats.py]: and ``print `` with ``print()`` for our Python 3 users

File: scipy/stats/morestats.py [hunk 9]
```diff
@@ -249,8 +275,25 @@ def kstatvar(data, n=2):
...
+    var(k_{1}) = \frac{\kappa^2}{n}
+    var(k_{2}) = \frac{\kappa^4}{n} + \frac{2\kappa^2_{2}}{n - 1}
+    var(k_{3}) = \frac{\kappa^6}{n} + \frac{9 \kappa_2 \kappa_4}{n - 1} +
+                 \frac{9 \kappa^2_{3}}{n - 1} +
+                 \frac{6 n \kappa^3_{2}}{(n-1) (n-2)}
+    var(k_{4}) = \frac{\kappa^8}{n} + \frac{16 \kappa_2 \kappa_6}{n - 1} +
+                 \frac{48 \kappa_{3} \kappa_5}{n - 1} +
+                 \frac{34 \kappa^2_{4}}{n-1} + \frac{72 n \kappa^2_{2} \kappa_4}{(n - 1) (n - 2)} +
+                 \frac{144 n \kappa_{2} \kappa^2_{3}}{(n - 1) (n - 2)} +
+                 \frac{24 (n + 1) n \kappa^4_{2}}{(n - 1) (n - 2) (n - 3)}
```
Comment (rgommers) [inline scipy/stats/morestats.py]: The refguide check failed. This LaTeX text all needs a 4-space indent.

**Comment by rgommers** [comment]
Looks like this can be merged once TravisCI is happy.

**Comment by aeklant** [comment]
Alright, TravisCI is green now. There were indeed some formatting issues.

**Comment by rgommers** [comment]
LGTM now, merged. Thanks @aeklant @ev-br

**Pull request merged**